src/Castle.Windsor/Core/Internal/TypeUtil.cs

```diff
@@ -22,7 +22,20 @@ namespace Castle.Core.Internal
 	public static class TypeUtil
 	{
 		/// <summary>
-		///   Checkis if given <paramref name="type" /> is a primitive type or collection of primitive types. Value types, <see cref="string" /> are considered primitive and can not be registered as components in Windsor
+		/// Indicates whether the type is explicitly marked as nullable. Currently only detects nullable value types via
+		/// <see cref="Nullable{T}"/>, but could be expanded to detect C# nullable reference types.
+		/// </summary>
+		internal static bool IsNullable(this Type type)
+		{
+			// For now, only detects nullable value types. Detecting nullable reference types is complicated.
+			// Spec: https://github.com/dotnet/roslyn/blob/version-3.2.0/docs/features/nullable-metadata.md
+			// Reflection API proposal: https://github.com/dotnet/runtime/issues/29723
+
+			return type.GetTypeInfo().IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>);
```

Comment:
Any reason not to just use `Nullable.GetUnderlyingType(type) != null` inline?

Reply:
I guess this is a premature optimization (avoids accessing the generic type arguments array). Castle.Core implements `IsNullableType()` this way though.

Reply:
Ah, as far as the inline part. I was hoping that one day, nullable reference types would also be detected by this extension method.

Reply:
Haha, looks like I added `IsNullableType` to Castle Core two years ago after refactoring the same reflection check obviously without even a second thought.

I'll leave it up for you to decide, but I just thought you'd then just do it inline without this extra method, because nullable reference types are the same `System.Type` with attributes applied to the `ParameterInfo` so wouldn't be done in this method.

Reply:
I would assume this method would begin requiring the ParameterInfo to be passed at that point and almost wrote this in the comment :D Sadly, the attributes are sometimes not applied to the ParameterInfo but rather to the method or type for compression purposes. But my thought was that passing a second ICustomAttributeProvider parameter would allow the right traversals to take place.

I'll keep this since that's okay with you and I feel like it could increase the chances of nullable reference types being recognized.